pkg/rbd/nodeserver.go

```diff
@@ -148,24 +163,40 @@ func (ns *NodeServer) NodeStageVolume(ctx context.Context, req *csi.NodeStageVol
 	return &csi.NodeStageVolumeResponse{}, nil
 }
 
-func (ns *NodeServer) cleanupStagingPath(stagingTargetPath, devicePath, volID string, isStagePathCreated, isBlock, isMounted bool) {
+func (ns *NodeServer) undoStagingTransaction(stagingParentPath, devicePath, volID string, isStagePathCreated, isMounted bool) {
 	var err error
+
+	stagingTargetPath := stagingParentPath + "/" + volID
 	if isMounted {
 		err = ns.mounter.Unmount(stagingTargetPath)
 		if err != nil {
 			klog.Errorf("failed to unmount stagingtargetPath: %s with error: %v", stagingTargetPath, err)
+			return
 		}
 	}
-	// remove the block file created on staging path
-	if isBlock && isStagePathCreated {
+
+	// remove the file/directory created on staging path
+	if isStagePathCreated {
 		err = os.Remove(stagingTargetPath)
 		if err != nil {
 			klog.Errorf("failed to remove stagingtargetPath: %s with error: %v", stagingTargetPath, err)
+			// continue on failure to unmap the image, as leaving stale images causes more issues than a stale file/directory
 		}
 	}
+
 	// Unmapping rbd device
-	if err = detachRBDDevice(devicePath); err != nil {
-		klog.Errorf("failed to unmap rbd device: %s for volume %s with error: %v", devicePath, volID, err)
+	if devicePath != "" {
+		err = detachRBDDevice(devicePath)
+		if err != nil {
+			klog.Errorf("failed to unmap rbd device: %s for volume %s with error: %v", devicePath, volID, err)
+			return
```

Comment:
if detach fails do we need to leave the image metadata file?

Reply:
Strictly we do not expect a NodeUnstage call (which is where the file is used) as Stage failed or has not returned in time. This is retained to maintain the order of operations and not partially cleanup the state created.

Reply:
I would like to get it fixed, this will leave a stale file and directory, as Kube will fail to delete staging path

Reply:
fixed.